src/ReverseProxy/Service/Management/ProxyConfigManager.cs

```diff
@@ -3,66 +3,266 @@
         private readonly IRuntimeRouteBuilder _routeEndpointBuilder;
         private readonly IClusterManager _clusterManager;
         private readonly IRouteManager _routeManager;
-        private readonly IProxyDynamicEndpointDataSource _dynamicEndpointDataSource;
+        private readonly IEnumerable<IProxyConfigFilter> _filters;
+        private readonly IConfigValidator _configValidator;
+        private IDisposable _changeSubscription;
 
-        public ReverseProxyConfigManager(
-            ILogger<ReverseProxyConfigManager> logger,
-            IDynamicConfigBuilder configBuilder,
+        private List<Endpoint> _endpoints = new List<Endpoint>(0);
+        private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+        private IChangeToken _changeToken;
+
+        public ProxyConfigManager(
+            ILogger<ProxyConfigManager> logger,
+            IProxyConfigProvider provider,
             IRuntimeRouteBuilder routeEndpointBuilder,
             IClusterManager clusterManager,
             IRouteManager routeManager,
-            IProxyDynamicEndpointDataSource dynamicEndpointDataSource)
+            IEnumerable<IProxyConfigFilter> filters,
+            IConfigValidator configValidator)
         {
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
-            _configBuilder = configBuilder ?? throw new ArgumentNullException(nameof(configBuilder));
+            _provider = provider ?? throw new ArgumentNullException(nameof(provider));
             _routeEndpointBuilder = routeEndpointBuilder ?? throw new ArgumentNullException(nameof(routeEndpointBuilder));
             _clusterManager = clusterManager ?? throw new ArgumentNullException(nameof(clusterManager));
             _routeManager = routeManager ?? throw new ArgumentNullException(nameof(routeManager));
-            _dynamicEndpointDataSource = dynamicEndpointDataSource ?? throw new ArgumentNullException(nameof(dynamicEndpointDataSource));
+            _filters = filters ?? throw new ArgumentNullException(nameof(filters));
+            _configValidator = configValidator ?? throw new ArgumentNullException(nameof(configValidator));
+
+            _changeToken = new CancellationChangeToken(_cancellationTokenSource.Token);
         }
 
+        // EndpointDataSource
+
+        /// <inheritdoc/>
+        public override IReadOnlyList<Endpoint> Endpoints => Volatile.Read(ref _endpoints);
+
+        /// <inheritdoc/>
+        public override IChangeToken GetChangeToken() => Volatile.Read(ref _changeToken);
+
+        // IProxyConfigManager
+
         /// <inheritdoc/>
-        public async Task ApplyConfigurationsAsync(CancellationToken cancellation)
+        public async Task<EndpointDataSource> InitialLoadAsync()
+        {
+            // Trigger the first load immediately and throw if it fails.
+            // We intend this to crash the app so we don't try listening for further changes.
```

Comment:
"listening for" -> "listening to"?
```suggestion
            // We intend this to crash the app so we don't try listening to further changes.
```